diff --color -ruN orig/googletest/include/gtest/gtest-typed-test.h patched/googletest/include/gtest/gtest-typed-test.h
--- orig/googletest/include/gtest/gtest-typed-test.h    2024-05-28 08:42:39
+++ patched/googletest/include/gtest/gtest-typed-test.h 2024-06-04 14:10:13
@@ -200,10 +200,11 @@
   template <typename gtest_TypeParam_>                                       \
   class GTEST_TEST_CLASS_NAME_(CaseName, TestName)                           \
       : public CaseName<gtest_TypeParam_> {                                  \
+   public:                                                                   \
+    void TestBody() override;                                                \
    private:                                                                  \
     typedef CaseName<gtest_TypeParam_> TestFixture;                          \
     typedef gtest_TypeParam_ TypeParam;                                      \
-    void TestBody() override;                                                \
   };                                                                         \
   GTEST_INTERNAL_ATTRIBUTE_MAYBE_UNUSED static bool                          \
       gtest_##CaseName##_##TestName##_registered_ =                          \
@@ -272,10 +273,11 @@
   namespace GTEST_SUITE_NAMESPACE_(SuiteName) {                   \
   template <typename gtest_TypeParam_>                            \
   class TestName : public SuiteName<gtest_TypeParam_> {           \
+   public:                                                        \
+    void TestBody() override;                                     \
    private:                                                       \
     typedef SuiteName<gtest_TypeParam_> TestFixture;              \
     typedef gtest_TypeParam_ TypeParam;                           \
-    void TestBody() override;                                     \
   };                                                              \
   GTEST_INTERNAL_ATTRIBUTE_MAYBE_UNUSED static bool               \
       gtest_##TestName##_defined_ =                               \
diff --color -ruN orig/googletest/include/gtest/internal/gtest-internal.h patched/googletest/include/gtest/internal/gtest-internal.h
--- orig/googletest/include/gtest/internal/gtest-internal.h     2024-05-28 08:42:39
+++ patched/googletest/include/gtest/internal/gtest-internal.h  2024-06-04 14:10:48
@@ -1499,8 +1499,9 @@
         GTEST_TEST_CLASS_NAME_(test_suite_name,                                \
                                test_name) &&) noexcept = delete; /* NOLINT */  \
                                                                                \
-   private:                                                                    \
+   public:                                                                     \
     void TestBody() override;                                                  \
+   private:                                                                    \
     GTEST_INTERNAL_ATTRIBUTE_MAYBE_UNUSED static ::testing::TestInfo* const    \
         test_info_;                                                            \
   };                                                                           \
